text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,379 @@
+            // Don't print incomplete text
+            return;
+        }
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << std::flush;
+        print_len = text.size();
+    }
+
+    void end() {
+        std::string text = detokenize(detokenizer, token_cache);
+        std::cout << std::string_view{text.data() + print_len, text.size() - print_len} << '\n';
+        token_cache.clear();
+        print_len = 0;
+    }
+};
+}
+
+std::vector<int64_t> convert_to_vector(ov::Tensor tensor) {
+    std::vector<int64_t> res_vector;
+    for (int i = 0; i << tensor.get_shape().back(); i++) {
+        res_vector.emplace_back(tensor.data<int64_t>()[i]);
+    }
+    return res_vector;
+}
+
+ov::Tensor append_element(ov::Tensor tensor_val, int64_t element) {
+    auto vec = convert_to_vector(tensor_val);
+    vec.emplace_back(element);
+    return ov::Tensor{ov::element::i64, {BATCH_SIZE, tensor_val.get_shape().back()}, vec.data()};
+}
+
+void set_key_values(ov::InferRequest& request, int size) {
+    // forwards key values from Results to inputs
+    std::stringstream ss_1, ss_2;
+
+    for (int i = 0; i < size; i++) {
+        ss_1 << "present." << i << ".key";
+        ss_2 << "past_key_values." << i << ".key";
+        request.set_tensor(ss_2.str(), request.get_tensor(ss_1.str()));
+        ss_1.str("");
+        ss_2.str("");
+
+        ss_1 << "present." << i << ".value";
+        ss_2 << "past_key_values." << i << ".value";
+        request.set_tensor(ss_2.str(), request.get_tensor(ss_1.str()));
+        ss_1.str("");
+        ss_2.str("");
+    }
+}
+
+void init_key_values(ov::InferRequest request, int kv_length) {
+    std::stringstream ss;
+    ss.clear();
+    
+    for (int i = 0; i < kv_length; i++) {
+        ss << "past_key_values." << i << ".key";
+        auto tensor_shape = request.get_tensor(ss.str()).get_shape();
+
+        request.set_tensor(ss.str(), ov::Tensor(ov::element::f32, {BATCH_SIZE, tensor_shape[1], 0, tensor_shape[3]}));
+        ss.str("");
+
+        ss << "past_key_values." << i << ".value";
+        request.set_tensor(ss.str(), ov::Tensor(ov::element::f32, {BATCH_SIZE, tensor_shape[1], 0, tensor_shape[3]}));
+        ss.str("");
+    }
+}
+
+ov::Tensor trimm_tensor(ov::Tensor tensor, uint64_t axis, uint64_t new_seq_len) {
+    // tensor shape = [batch, size_1, seq_len, size_2]
+    auto old_tensor_data = tensor.data<float>();
+    auto shape = tensor.get_shape();
+    auto size_1 = shape[1];
+    auto old_seq_len = shape[2];
+    auto size_2 = shape[3];
+    
+    // todo: assume batch is always 1, if different one more cycle is needed
+    // copy elements from the old to the new tensor
+    auto new_tensor = ov::Tensor{ov::element::f32, {BATCH_SIZE, size_1, new_seq_len, size_2}};
+    auto new_tensor_data = new_tensor.data<float>();
+    for (int i = 0; i < size_1; i++) {
+        for (int j = 0; j < new_seq_len; j++) {
+            for (int k = 0; k < size_2; k++) {
```

Comment:
we can use std::memcpy instead of inner loop

Reply:
sure, memcpy should be better. Also thought about that

Reply:
done. In this place I will also add one more cycle for batched, just to be sure.